Source/HTTP/iOS/http_ios.mm

```diff
@@ -0,0 +1,134 @@
+        const char *valueCString = [value cStringUsingEncoding:NSUTF8StringEncoding];
+        HCHttpCallResponseSetHeader(m_call, keyCString, valueCString);
+    }
+
+    HCHttpCallResponseSetResponseBodyBytes(m_call, static_cast<const uint8_t*>([data bytes]), [data length]);
+    CompleteAsync(m_asyncBlock, S_OK, 0);
+}
+
+void ios_http::initiate_request()
+{
+    const char* urlCString = nullptr;
+    const char* methodCString = nullptr;
+    if (FAILED(HCHttpCallRequestGetUrl(m_call, &methodCString, &urlCString)))
+    {
+        CompleteAsync(m_asyncBlock, E_FAIL, 0);
+        return; //TODO: handle errors better
+    }
+    
+    NSString* urlString = [[NSString alloc] initWithUTF8String:urlCString];
+    NSURL* url = [NSURL URLWithString:urlString];
+    
+    NSString* methodString = [[NSString alloc] initWithUTF8String:methodCString];
+    
+    NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL:url];
+    [request setHTTPMethod:methodString];
+    
+    uint32_t numHeaders = 0;
+    if (FAILED(HCHttpCallRequestGetNumHeaders(m_call, &numHeaders)))
+    {
+        CompleteAsync(m_asyncBlock, E_FAIL, 0);
+        return; //TODO: errors
+    }
+    
+    for(uint32_t i=0; i<numHeaders; ++i)
+    {
+        const char* headerName;
+        const char* headerValue;
+        if (SUCCEEDED(HCHttpCallRequestGetHeaderAtIndex(m_call, i, &headerName, &headerValue)))
+        {
+            NSString* headerNameString = [[NSString alloc] initWithUTF8String:headerName];
+            NSString* headerValueString = [[NSString alloc] initWithUTF8String:headerValue];
+            
+            [request addValue:headerValueString forHTTPHeaderField:headerNameString];
+        }
+    }
+    
+    const BYTE* requestBody = nullptr;
+    uint32_t requestBodySize = 0;
+    if(FAILED(HCHttpCallRequestGetRequestBodyBytes(m_call, &requestBody, &requestBodySize)))
+    {
+        CompleteAsync(m_asyncBlock, E_FAIL, 0);
+        return; //TODO: errors
+    }
+    
+    if (requestBodySize == 0)
+    {
+        m_sessionTask = [m_session dataTaskWithRequest:request completionHandler:
+                         ^(NSData *data, NSURLResponse *response, NSError *error)
+                         {
+                             this->completion_handler(data, response, error);
+                         }];
+    }
+    else
+    {
+        NSData* data = [NSData dataWithBytes:requestBody length:requestBodySize];
+        
+        m_sessionTask = [m_session uploadTaskWithRequest:request fromData:data completionHandler:
+                         ^(NSData *data, NSURLResponse *response, NSError *error)
+                         {
+                             this->completion_handler(data, response, error);
+                         }];
+    }
+    
+    [m_sessionTask resume];
+}
+
+
+
+void Internal_HCHttpCallPerform(
+    _In_ AsyncBlock* asyncBlock,
+    _In_ hc_call_handle_t call
+    )
+{
+    ios_http* httpTask = new ios_http(asyncBlock, call);
```

Comment:
how is httpTask freed?

Reply:
Side comment: as a way to annotate to ourselves that we're in charge of this memory, make httpTask a smart pointer, and then add a release() call at the end of the method. This makes it clear to the reader that it will be leaked if not handled.

---
In reply to: [188699524](https://github.com/Microsoft/libHttpClient/pull/162#discussion_r188699524) [](ancestors = 188699524)

Reply:
Now we just need to fix the leak.
You should wrap this in a unique_ptr first thing in each of the callbacks so that it is destroyed when we leave the callback

---
In reply to: [188699524](https://github.com/Microsoft/libHttpClient/pull/162#discussion_r188699524) [](ancestors = 188699524)